base/setup/lib/utils/filesup.c

```diff
@@ -71,78 +68,84 @@ SetupCreateSingleDirectory(
                           NULL,
                           0);
     if (NT_SUCCESS(Status))
-    {
         NtClose(DirectoryHandle);
-    }
-
-    RtlFreeUnicodeString(&PathName);
 
     return Status;
 }
 
 NTSTATUS
 SetupCreateDirectory(
-    IN PCWSTR PathName)
+    _In_ PCWSTR DirectoryName)
 {
     NTSTATUS Status = STATUS_SUCCESS;
-    PWCHAR PathBuffer = NULL;
+    UNICODE_STRING PathNameU;
+    PWCHAR PathBuffer;
     PWCHAR Ptr, EndPtr;
-    ULONG BackslashCount;
-    ULONG Size;
-
-    Size = (wcslen(PathName) + 1) * sizeof(WCHAR);
-    PathBuffer = RtlAllocateHeap(RtlGetProcessHeap(), HEAP_ZERO_MEMORY, Size);
-    if (PathBuffer == NULL)
-        return STATUS_INSUFFICIENT_RESOURCES;
 
-    wcscpy(PathBuffer, PathName);
-    EndPtr = PathBuffer + wcslen(PathName);
-
-    Ptr = PathBuffer;
+    if (!RtlCreateUnicodeString(&PathNameU, DirectoryName))
+        return STATUS_NO_MEMORY;
+    PathBuffer = PathNameU.Buffer;
+    EndPtr = PathBuffer + (PathNameU.Length / sizeof(WCHAR));
 
-    /* Skip the '\Device\HarddiskX\PartitionY\ part */
-    BackslashCount = 0;
-    while (Ptr < EndPtr && BackslashCount < 4)
+    /* Find the deepest existing sub-directory: start from the
+     * end and go back verifying each sub-directory in turn */
+    for (Ptr = EndPtr; PathBuffer < Ptr;)
```

Comment:
nit: It would make more sense to reverse the condition imo (`Ptr > PathBuffer`)

Reply:
done